오픈 마켓 [STEP 2] 예거, 릴리

Jager-yoo opened the pull request:
안녕하세요! 찰리 ~! 😄 @kcharliek 

`STEP 2` 내용에 대해 PR 올립니다!
더 수정할 부분이 남았지만, 리뷰 받으면서 점차 고쳐나가려고 합니다.
많이 배우고 있습니다. 앞으로 2주 동안에도 잘 부탁드립니다. 🙏🏻 

# 📱 앱 시연 영상

https://user-images.githubusercontent.com/71127966/149471453-0cc7f16f-95b4-46e5-9545-16fe99a86cca.mov

<br>

# 🛠️ 구현 내용

## STEP 2

- `스토리보드`를 사용해서 UI 컴포넌트를 배치하고 `오토레이아웃`을 적용하고 UI 컴포넌트들의 디테일한 속성(폰트, 색상 등)은 `코드`를 통해 구현했습니다.
  - `Navigation Controller`를 임베드해서 나온 `Navigation Bar` 위에 `Segmented Control`을 올려서, 모든 화면이 해당 UI 컴포넌트를 공유할 수 있게 만들었습니다.

- `Tab Bar Controller`를 임베드해서 2개의 Tab Bar Item 을 각각의 화면(테이블, 컬렉션)으로 연결했습니다.
  - tabBar 는 `hidden` 처리하고 `Segmented Control`로만 화면을 변경할 수 있게 만들었습니다.

- 앱 실행 시 보여야 하는 `상품 리스트`를 가져오는 통신은 viewDidLoad() 에서 요청하고 그 결과는 미리 만들어둔 프로퍼티에 넣어줬습니다.
  - 통신은 `비동기`적으로 진행돼서 완료 시점을 정확하게 알 수 없으므로, 완료되면 `DispatchQueue.main.async`로 `reloadData()` 메서드를 호출해서 화면에 다시 데이터를 그리도록 했습니다.
  - IBOutlet 으로 연결한 커스텀 셀에는 통신 결과를 `UIImage`, `NSAttributedString` 타입으로 변경해서 값을 할당해줬습니다.

<br>

# 🔥 고민한 내용

### 1️⃣ 스토리보드와 programmatic UI 를 조합하는 것에 대한 고민

이번 스텝2 진행시간이 넉넉하지 않아 스토리보드와 코드를 섞어서 UI를 구성해보았습니다.
각 각의 방법의 장점을 취해서 쓴다면 좋을 거라 생각했습니다.

#### 🖼️ 스토리보드 장점 
- 코드량을 줄일 수 있음
- 가시적으로 확인 가능

#### ⌨️ 코드 구현 장점
- 스토리보드에서 잡기 힘든 부분 까지 미세하게 설정 가능

<br>

### 2️⃣ 첫 리스트 조회 API호출을 어디에서 해줄지

처음 `ProductListViewController` 와 `ProductGridViewController` 로 화면을 불러올 때,
테이블 뷰와 컬렉션 뷰를 그려주기 위해 **상품리스트조회 API**를 호출해서 응답으로 받은 정보를 사용해야합니다.

즉 동일 API호출의 Response를 공유하고 있다고 판단했습니다.

그래서 API호출을 한 번만 해서 응답에 대한 데이터를 두 뷰 컨트롤러간에 공유할 수 있는 방법을 고민해봤습니다.

- TabBarController에서 API를 호출한 뒤, 데이터를 저장, selectedIndex를 활용해서 각 뷰컨틀롤러에 전달하는 방법

테이블 뷰와 컬렉션 뷰를 그려주는 것이 이번 스텝에서 더 우선되어야할 것 같아서, 위 방법에 대해서 좀 더 찾아보지는 못했습니다🥲 

우선 저희는 각 뷰컨트롤러의 `viewDidLoad()` 에서 API호출을 한 후 각 뷰컨의 프로퍼티에 저장해서 사용했습니다. 

저희가 생각한 방법외에 API호출을 한번만 하고 첫 응답의 데이터를 공유하는 좋은 방법이 있다면 힌트를 주시면 감사하겠습니다!

<br>

### 3️⃣ `제품명` 이 길어질 때 어떻게 처리할지

제품명이 길어지면 어떻게 보여주어야할 지 고민했습니다. 

우선 `잔여수량: xxx` label이 줄어들면 정보를 부정확하게 전달한다고 생각해서, `잔여수량: xxx` 를 content Compression Resistance를 1000 으로 설정해서 지켜주었습니다. 

그리고 고민되었던 것이 제품명을 줄일 것인지, 아니면 full Name으로 보여주게 할 것인지인데요~

저희가 생각해본 두가지 방법은 아래와 같습니다.

1. line을 1 로 고정 시켜서, 이름이 길어질 때 이름을 줄이는 방법 `또로로롤롱...` (✅ 우선 이 방법으로 구현했습니다!)
2. line을 0 으로 주고, 줄 바꿈을 하게되면 글자 사이즈를 자동으로 줄여서 full name이 보이도록 하는 방법

<p align="center"><img src="https://user-images.githubusercontent.com/71127966/149482992-11386417-eaff-4edb-afdd-97ac42866ced.png" width="60%"></p> 

2번 방법이 상품 이름에 대한 충분한 정보를 전달해주기에 더 좋다고 생각은 했지만, 
font size를 유동적으로 줄이는 방법에 대한 추가적인 공부가 필요할 것 같아 먼저 1번 방법으로 구현한 후 리팩터링시 고민해보자고 상의했습니다.

<br>

### 4️⃣ URL 이미지 다운로드 실패시 imageSymbol을 default 로 보여주게 구현

이미지 다운로드가 실패하면 비어있는 `UIImage()`를 리턴해줄까 고민하다가, 이미지가 빈 칸으로 주어지는 것 보다는, 다운로드가 실패한 듯한 늬앙스가 담긴 로고를 보여주고 싶었습니다.

그래서 다운로드가 실패하면, 빈 칸이 아니라 "xmark.icloud" 아이콘을 노출시키도록 에러 처리 화면을 만들었습니다.

![image](https://user-images.githubusercontent.com/71127966/149483089-51f8a7a3-2ea1-497e-bad7-624280c04f3f.png)

<br>

# 🤦🏻‍♂️ 아직 구현하지 못한 부분

- Loading Indicator
- 오류 발생시 Alert 띄우도록 구현
- 잔여 수량의 자리수에 따라 `잔여수량: XXX` 의 leading 이 불규칙하게 나열되는 모습 손보기
- GridLayout 에서 가격 표시가 좌, 우가 아니라 위, 아래로 배치되도록 만들기

<br>

# ❓ 궁금한 부분

### 1️⃣ 현업에서는 스토리보드와 코드로 만드는 UI를 섞어서 사용하기도 하나요? 둘을 함께 사용하는 것에 대한 찰리의 의견이 궁금합니다!

<br>

Jager-yoo committed:
feat: TabBarController 임베드하고 TableViewController, CollectionViewController 연결

Jager-yoo committed:
fix: 삭제된 ViewController 2개 복구

yeahg-dev committed:
refactor: excute메서드의 Result 연관타입을 JSON디코딩해서 받을 수 있도록 수정

Jager-yoo committed:
feat: Navigation Bar 위에 커스텀 Segmented Control 구현

yeahg-dev committed:
feat: switchViewController 메서드 구현

Jager-yoo committed:
feat: 스토리보드에서 커스텀 ProductListViewCell 구현하고 IBOutlet 연결

yeahg-dev committed:
feat: 스토리보드에서 커스텀 ProductGridViewCell 구현하고 IBOutlet 연결

Jager-yoo committed:
feat: 2가지 커스텀 셀의 reuseIdentifier 프로퍼티 구현

yeahg-dev committed:
refactor: executeTask 메서드 리팩터링

yeahg-dev committed:
refactor: 커스텀셀 constraint 리팩터링 및 접근제어 수정

Jager-yoo committed:
feat: 커스텀 ProductListViewCell 의 3가지 UILabel에 넣을 attributedString 구현

- Product 타입의 extension 만들어서 프로퍼티 정보를 조합하여 NSAttributedString 타입으로 반환해주는 연산 프로퍼티 3개 구현
- 하드코딩 최소화를 위해 Double, String, NSAttributedString 타입의 extension 구현
- 스토리보드의 '잔여 수량' 오른쪽 버튼과의 constraint 10pt 부여

yeahg-dev committed:
feat: getImage 메서드 구현 및 productListViewCell의 image 세팅 코드 추가

Jager-yoo committed:
feat: ProductGridViewController 구현 중간 저장

yeahg-dev committed:
fix: CollectionView FlowLayout 설정해서 item이2열로 보이도록 수정

Jager-yoo committed:
feat: ProductGridViewCell 레이아웃 수정

yeahg-dev committed:
file: Extension 폴더 Network 폴더밖으로 이동

kcharliek commented:
## 고민하신 내용에 대해

1. 스토리보드와 programmatic UI 를 조합하는 것

스토리보드 만으로 모든 UI를 개발하기는 쉽지 않습니다. 당연한 것이라고 생각합니다.

2.  첫 리스트 조회 API호출을 어디에서 해줄지

> API호출을 한 번만 해서 응답에 대한 데이터를 두 뷰 컨트롤러간에 공유할 수 있는 방법을 고민

이라고 적어주셨는데요. 아주 바람직한 방향입니다. 여러번 호출해서 네트워크 낭비할 필요없죠.
저는 class를 사용해서, reference를 이용하는 것을 제안드릴게요. 방법은 class DataStore{} 을 정의해서 각 viewController에 주입해주는 것입니다. DataLoad와 데이터 Holding을 모두 DataStore가 하고요. 각 viewController는 reference로 DataStore을 들고있으니, 바라보는 데이터는 모두 같을 것입니다. 데이터 변화를 각 VC가 어떻게 모니터링 할지는 다음 고민이지만, 일단은 DataStore를 두면 데이터 공유 문제는 해결될 것 입니다.

3. 제품명 이 길어질 때 어떻게 처리할지

minimum font size를 이용할 수도 있고요. marqueelabel 을 구현할 수도 있고요. 방법은 많은데 정답은 없어서 괜찮아 보이는 것을 선택하시면 좋을 것 같습니다.


4. URL 이미지 다운로드 실패시 imageSymbol을 default 로 보여주게 구현

좋은 아이디어라고 생각합니다. 잘 해주셨어요 :)

kcharliek commented:
## 궁금한 부분에 대해

> 현업에서는 스토리보드와 코드로 만드는 UI를 섞어서 사용하기도 하나요? 둘을 함께 사용하는 것에 대한 의견 바랍니다

코드로만 UI를 구현할 수도 있지만, 스토리보드(인터페이스 빌더) 를 사용해서 UI 모두를 구현하기는 어렵습니다. 섞어서 사용하는 것에 대해 이상해보이지 않습니다.

kcharliek commented on OpenMarket/OpenMarket/Cell/ProductGridViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+
+import UIKit
+
+class ProductGridViewCell: UICollectionViewCell {
```
iOS 의 공통 컨벤션인데요. Class명에 Component 명을 포함시킵니다.
UICollectionViewCell 의 subclass 는 XXXCollectionViewCell
UITextVIew의 subclass는 XXXTextView 와 같이 표기하는 것이 일반적입니다.

그래서 위 경우 ProductGridCollectionViewCell 이라고 네이밍 하는 것이 좋아보입니다.

yeahg-dev replied:
ViewController와 TableViewCell, CollectionViewCell모두 UIComponent명을 포함시키는 네이밍으로 수정했습니다!

리스트와 그리드 대신 테이블, 컬렉션이라는 용어를 사용해서, ViewController의 네이밍도 통일성있게 수정했습니다😸

kcharliek commented on OpenMarket/OpenMarket/Cell/ProductGridViewCell.swift:
```diff
@@ -0,0 +1,18 @@
+
+class ProductGridViewCell: UICollectionViewCell {
+    
+    static let reuseIdentifier = String(describing: ProductGridViewCell.self)
```
ProductListViewCell, ProductGridViewCell 에서 공통적으로 사용하고 있는 것 같은데요. UICollectionViewCell extension으로 공통 처리할 수 있을 것 같습니다.

Jager-yoo replied:
`ProductListViewCell` 클래스는 `UITableViewCell` 을 상속받고 있어서 `UICollectionViewCell` 의 extension 으로 공통적으로 처리하기 어려울 것 같습니다! 😅 

다른 방법을 좀 더 고민해보겠습니다!

kcharliek replied:
그렇군요. 그렇다면 'protocol Reusable {}' 과 같은 프로토콜을 두어 해결해볼 수 있을 것 같습니다.

Jager-yoo replied:
아래에서 찰리가 말씀해주신 `dequeue 메서드 extension` 을 적용했더니, 각 커스텀 뷰셀 타입이 `reuseIdentifier 프로퍼티`를 소유할 필요가 없어졌습니다! 😅
의도하진 않았지만, 관리해야할 프로퍼티가 사라졌기 때문에, 문제가 해결됐네요 ㅎㅎㅎ

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductGridViewController.swift:
```diff
@@ -0,0 +1,96 @@
+        _ collectionView: UICollectionView,
+        cellForItemAt indexPath: IndexPath
+    ) -> UICollectionViewCell {
+        guard let cell = collectionView.dequeueReusableCell(
```
지금 작성해주신 dequeue 방법도 괜찮은데요. 현업에서는 아래와 같은 extension 형태로 많이 사용합니다. 아래와 같이 사용했을 때의 장점은 reuseIdentifier를 관리할 필요가 없다는 점 입니다. 정말 많이 사용하는 extension이라 추천드립니다.

``` swift    
func dequeueReusableCell<T: UICollectionViewCell>(withClass name: T.Type, for indexPath: IndexPath) -> T {
        guard let cell = dequeueReusableCell(withReuseIdentifier: String(describing: name), for: indexPath) as? T else {
            fatalError(
                "Couldn't find UICollectionViewCell for \(String(describing: name)), make sure the cell is registered with collection view")
        }
        return cell
    }
```

ref: https://github.com/SwifterSwift/SwifterSwift/blob/master/Sources/SwifterSwift/UIKit/UICollectionViewExtensions.swift

Jager-yoo replied:
좋은 extension 알려주셔서 감사합니다! 😄 바로 적용해봤어요~!

저희가 테이블 뷰컨과 컬렉션 뷰컨을 각각 사용하고 있기 때문에, 하나의 extension 으로 통합하진 못했고
뷰컨 파일의 하단에 각각 `private extension` 으로 메서드를 구현해두었습니다!

해당 메서드를 적용하면서 생긴 변화를 정리해봤습니다.
- 커스텀 Cell 타입의 네이밍을 그대로 받아 reuseIdentifier 로 지정해주고 타입 캐스팅까지 해준 뒤 옵셔널 바인딩이 완료된 상태의 'cell' 을 반환해주기 때문에 구현부의 코드가 깔끔해졌습니다.
- 또한, 각 커스텀 Cell 타입 내부에서 reuseIdentifier 프로퍼티를 관리할 필요가 사라졌어요!

그리고 질문이 있습니다. 🙋🏻‍♂️ 

해당 메서드를 저희가 구현한 것처럼, 각 뷰컨의 하단에 `private extension`으로서 놔두는 게 좋을지
아니면 `extension` 폴더의 별도 파일로 분리해두는 게 좋을지 고민이 됐습니다.

찰리였다면 어떻게 하셨을지 의견 주시면 감사하겠습니다!! 👍🏻

kcharliek replied:
해당 메서드는 비즈니스 로직이 없는 순수 '유틸' 의 기능을 하고 있으므로 별도 extension file로 분리할 것 같습니다.

yeahg-dev replied:
오호~ 네!! 비지니스 로직이 없는 순수 유틸 기능의 메서드인것 같아요! 분리했습니다😊

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductGridViewController.swift:
```diff
@@ -0,0 +1,96 @@
+            withReuseIdentifier: ProductGridViewCell.reuseIdentifier,
+            for: indexPath
+        ) as? ProductGridViewCell else {
+            return ProductGridViewCell()
```
1. UICollectionViewCell() (ProductGridViewCell()) 은 크래시를 유발합니다.

2. else문을 타는 경우는 reuseIdentifier가 등록되지 않은 경우인데요. 이 경우는 개발자 테스트 단계에서 무조건 걸러지기 때문에 else문을 타는 경우는 없다고 봐도 무방합니다. 

3. 제가 위에 알려드린 extension을 사용한다면 fatalError로 optional 처리가 되는데요. 현재 방법이라면  empty cell을 등록하는 식으로 구현해서 리턴해야할 것 같습니다.

Jager-yoo replied:
찰리가 알려주신 extension 을 적용해서 리팩토링했습니다! 👍🏻

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductGridViewController.swift:
```diff
@@ -0,0 +1,96 @@
+            return cell
+        }
+        
+        configureGridContent(of: cell, with: product)
```
모든 UI 컴포넌트는 private으로 선언하는 것이 기본이라고 생각하시면 좋습니다. 자신의 UI는 자신만 수정하도록 구현하는 것이 사이드 이펙트를 줄이는 방법이기 떄문입니다.

그렇다면 configureGridContent  같은 메서드를 구현하지 못하게 될 것인데요. 그렇다면 어떻게 구현하면 될지 고민 하고 수정해주시기 바랍니다.

yeahg-dev replied:
ProductCollectionViewCell에 product정보를 받아서 IBOutlet의 속성을 변경시키는 메서드를 만들어 주어,
타입의 프로퍼티는 스스로만 변경할 수 있도록 수정해주었습니다. 
그리고 해당메서드의 접근제어를 internal로 지정해주어 다른 객체에서 사용할 수 있도록 인터페이스를 제공해주도록 했습니다.

```swift
class ProductCollectionViewCell: UICollectionViewCell {
    
    @IBOutlet private weak var productThumbnail: UIImageView!
    @IBOutlet private weak var productName: UILabel!
    @IBOutlet private weak var productPrice: UILabel!
    @IBOutlet private weak var productStock: UILabel!
    
    func configureCollectionContent(with product: Product) {
        DispatchQueue.main.async {
            self.productThumbnail.image = self.getImage(from: product.thumbnail)
        }
        productName.attributedText = product.attributedName
        productPrice.attributedText = product.attributedPrice
        productStock.attributedText = product.attributedStock
    }
}
```

ProductTableViewCell도 동일한 방향으로 수정해봤습니다.

kcharliek replied:
잘 해주셨습니다 :)

kcharliek commented on OpenMarket/OpenMarket/Network/APIExecutor.swift:
```diff
@@ -26,21 +24,24 @@ struct APIExecutor {
     }
     
-    private func executeTask(request: URLRequest, _ completion: Handler?) {
+    private func executeTask<T: Decodable>(request: URLRequest, _ completion: @escaping (Result<T, Error>) -> Void ) {
```
Data -> Decodable로 인터페이스 잘 바꿔주셨습니다 :)

kcharliek commented on OpenMarket/OpenMarket/Extension/NSAttributedString+Extension.swift:
```diff
@@ -0,0 +1,13 @@
+
+extension NSAttributedString {
+    
+    static let whiteSpace = NSAttributedString(string: .whiteSpace)
```
static  사용에는 신중하셔야합니다. 정말 그렇게 밖에 구현할 방법이 없을 때에만 사용해주세요.

위 경우 computedProperty로 대체 가능해보입니다.

Jager-yoo replied:
코멘트 감사합니다!
위 구현은 "NSAttributedString" 타입이 필요한 자리에서 쩜(.) 한 번으로 프로퍼티를 바로 꺼내쓰기 위한 것이었는데요!

static let 을 -> `static var` 로 바꾸는 정도는 괜찮은 걸까요? (아래 예시)

```swift
static var whiteSpace: NSAttributedString {
    return NSAttributedString(string: .whiteSpace)
}
```

근데 이렇게 연산 프로퍼티로 변경하게 되면, 오히려 해당 프로퍼티를 호출할 때마다 `연산`을 해야 하니까 static let 에 비해서 불필요한 비용을 더 지불하게 되는 게 아닐까? 하는 생각이 듭니다. 🤔 

만약 static 이 붙지 않은 `인스턴스 연산 프로퍼티`로 만들게 되면, 저희가 의도한 방식으로 사용할 수가 없게 되구요 😢 

어떤 식의 연산 프로퍼티 대체를 말씀하셨던 것인지 힌트를 주실 수 있을까요??

kcharliek replied:
제가 작성 의도를 잘못 이해했습니다. 저는 아래와 같은 스타일을 생각했는데요. 이렇게 했을 때 .whiteSpace로 접근하기는 어려워보입니다. 

``` swift
var whiteSpace: NSAttributedString {
    return NSAttributedString(string: .whiteSpace)
}
```

그런 목적이라면 이 정도의 static 은 괜찮아보입니다. Resolve하겠습니다.

kcharliek commented on OpenMarket/OpenMarket/Extension/Double+Extension.swift:
```diff
@@ -0,0 +1,19 @@
+
+extension Double {
+    
+    private static let numberFormatter = NumberFormatter()
```
1. static 사용하지 않았으면 합니다. NumberFormatter를 static 으로 생성해서 여러번 생성하지 않도록 하는 방향 좋습니다만, 이를 들고있는 역할을 Double이 하는 것은 좋아보이지 않습니다. 이 경우에는 여러번 생성하는 것이 스타일을 해치는 것보다 나을 것 같아서 formatted 가 생성하도록 해주시면 좋겠습니다.

2. 제가 생각하기에 formatted는 이 과제 만의 '비즈니스 로직' 입니다. 저는 재사용 가능한 Util이 아니라면 공통 extension으로 두는 것에 대해 좋게 생각하지 않는데요. 비즈니스 로직을 extension으로 구현하고 싶다면, 사용하는 파일 내에 private extension으로 구현해서 다른 곳에서 참조하지 못하도록 하는 것이 좋을 것 같습니다.

yeahg-dev replied:
저희는 비지니스 로직을 "모델단에서 처리한 raw한 값`(100000)`을 사용자에게 적절한 형태`(100,000)`로 표시되게하는 로직"이라고 이해를 해보았는데요. 

그렇다면 사용자에게 보여지지 않는 내부 로직인 String타입을 데이터로 변환하는 아래 익스텐션 같은 경우는 비지니스 로직으로 볼 수 없는 걸까요?
```swift
extension Data {
    
    mutating func append(_ string: String) {
        guard let data = string.data(using: .utf8) else {
            return
        }
        self.append(data)
    }
}
```

아니면... 저희가 비지니스 로직을 너무 좁게 이해하고 있는걸까요?🥲

kcharliek replied:
비즈니스로직은 릴리와 예거가 만들고있는 '이 앱 만의 로직' 이라고 생각하시면 됩니다. 예를들어, 아래의 formatted는 .decimal로 numberStyle을 정하는데요. decimal을 사용하겠다고 하는 것은 '이 앱만의 로직' 입니다. 다른 앱, 예를들어 제가 만드는 앱 에서는 이 로직이 필요가 없어요. 그런 점에서 Util로 보기는 어렵다는 의미였습니다.

Data Extension의 경우는요. 'utf8 인코딩을 한다' 라는 비즈니스 로직만 파라미터로 빼 내면, 유틸이 되어 비즈니스로직이 아니라고 할 수 있을 것 같습니다.

Jager-yoo replied:
메서드 내부에서 사용되는 `String.Encoding` 타입을 파라미터로 분리시켜서, `유틸화` 해봤습니다! 😄 

```swift
extension Data {
    
    mutating func append(_ string: String, using encodingStyle: String.Encoding) {
        guard let data = string.data(using: encodingStyle) else {
            return
        }
        self.append(data)
    }
}
```

kcharliek commented on OpenMarket/OpenMarket/Cell/ProductListViewCell.swift:
```diff
@@ -0,0 +1,30 @@
+    
+    static let reuseIdentifier = String(describing: ProductListViewCell.self)
+    
+    @IBOutlet weak var productThumbnail: UIImageView!
```
1. UI 컴포넌트는 private 으로 선언하는 것이 좋습니다. 접근 제어를 여는 것 말고는 더이상 다른 방법이 없을 때에만 하는 수 없이 풀어주세요.

2.  IBOutlet은 Optional (?) 로 선언해주세요. 보통 그럴일이 없고, 개발자들이 알아서 잘 챙기겠지만, 실수로 값이 nil이 되어 크래시나는 경우를 종종 보았습니다. 항상 안전 제일입니다.

yeahg-dev replied:
안전 제일! 동감합니다!! 👷 
!를 ?로 수정해주었습니다
프로퍼티를 사용하는 부분에서는 옵셔널 체이닝을 적용해주었습니다.

이 정도면 괜찮은 걸까요????

kcharliek replied:
넵 잘 해주셨습니다~!

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductGridViewController.swift:
```diff
@@ -0,0 +1,96 @@
+            return ProductGridViewCell()
+        }
+        
+        guard let product = initialProductsListPage?.pages[indexPath.item] else {
```
제가 담당했던 서비스들에서는 array indexing을 할 때 'index out of bounds' exception을 방지하기위해 subscript를 사용했습니다. 수정 필수는 아니고요, 참고만 해주세요 

https://stackoverflow.com/questions/25329186/safe-bounds-checked-array-lookup-in-swift-through-optional-bindings

yeahg-dev replied:
`index out of bounds` exception은 쉽게 일어날 수 있는 에러인 것 같아 별도 처리를 해주는 것이 안전하다!고 생각이 들었습니다.

저희가 실험을 해보았는데요👩🏻‍🔬
<img width="1398" alt="스크린샷 2022-01-17 오후 6 59 03" src="https://user-images.githubusercontent.com/81469717/149748717-ab342ac5-130d-4e69-a1e1-375ec5a9b47e.png">
위와 같이 range를 벗어난 index를 넣어주면 else문으로 빠지기도전에 Fatal Error가 발생하더라구요ㅠㅠ
그래서 참고주신 링크의 방법 처럼, range를 벗어난 index가 들어올 경우 `nil`이라도 반환시켜서 빈 cell을 반환하도록 수정해보겠습니다!

Jager-yoo replied:
배열(Array)에서 `index out of range`를 방지하기 위해, 찰리가 말씀해주신 `subscript(safe:)`를 Array 의 Extension 으로 구현했습니다!

링크에서는 조건문으로 삼항연산자를 사용하고 있는데요, 저희는 `switch문`을 사용해서, 좀 더 분명하게 로직을 구분해주었습니다.

```swift
extension Array {
    
    /// Returns the element at the specified index if it is within bounds, otherwise nil.
    subscript(safe index: Index) -> Element? {
        
        switch indices.contains(index) {
        case true:
            return self[index]
        case false:
            return nil
        }
    }
}

```

kcharliek commented on OpenMarket/OpenMarket/Controller/ProductGridViewController.swift:
```diff
@@ -0,0 +1,96 @@
+        super.viewDidLoad()
+        configureGridLayout()
+        
+        let request = ProductsListPageRequest(pageNo: 1, itemsPerPage: 20)
```
데이터를 가져오는 과정을 method로 한번 감싸주시면 가독성이 더 좋아질 것 같습니다.

yeahg-dev committed:
fix: productRegistration기능 디버깅

Jager-yoo committed:
refactor: 상품 thumbnail 이미지 다운로드 실패 시, 대신 보여줄 defaultImage 의 컬러를 systemGray 로 수정

Jager-yoo committed:
refactor: 상품 thumbnail 이미지 다운로드 및 ImageView 할당 작업이 비동기적으로 처리되도록 수정

yeahg-dev committed:
refactor: 뷰 관련 클래스를 UIComponent명을 포함시키는 네이밍으로 변경

Jager-yoo committed:
refactor: UITableView, UICollectionView 의 extension 으로 dequeueReusableCell 메서드 구현

- 커스텀 Cell 타입의 네이밍을 그대로 받아 reuseIdentifier 로 지정해주고 타입 캐스팅까지 해준 뒤 옵셔널 바인딩이 완료된 상태의 'cell' 을 반환해주는 메서드 구현
- 위 메서드를 적용하면 각 커스텀 Cell 타입 내부에서 reuseIdentifier 프로퍼티를 관리할 필요가 사라짐

yeahg-dev committed:
refactor: ProductCollectionViewCell의 IBOutlet을 타입 스스로 변경하도록 수정

Jager-yoo committed:
refactor: ProductTableViewCell 의 IBOutlet 을 타입 스스로 변경하도록 수정

yeahg-dev committed:
IBOutlet프로퍼티 옵셔널 타입 수정(! -> ?)

Jager-yoo committed:
'index out of range' fatal error 를 방지하기 위해 Array Extension 에 subscript(safe:) 구현

yeahg-dev commented:
안녕하세요 찰리!
리팩터링 도중 아래와 같은 문제를 발견해서, 이를 해결하기 위한 코드를 작성하다가 해결하지 못한 부분이 있어 질문드립니다!

## 빠르게 scroll시 cell의 이미지가 바뀌는 문제

스크롤을 빠르게 할 때, cell에 productName에 해당하는 이미지가 아닌 다른 이미지가 나타났다가 productName에 해당하는 이미지가 업데이트 되는 이슈가 있었습니다. 
그래서 이미지를 적용시켜주는 메서드에서 cell의 indexPath를 비교해서 동일할 때만 image가 업데이트 되도록 분기 처리해줘서 해결하고자 했습니다. 그런데 이미지가 아예 적용되지않는 이슈가 생겨 디버깅을 해보았더니 **`collectionView.indexPath(for: cell)`이 nil을 반환**해주고 있었어요..😨 구글링을 해보고 저희가 생각을 해보았는데.. 대체 왜 nil이 반환되는지 해결하지 못했습니다. 

저희 코드에서 nil이 반환되는 건 당연한건가요...? 

```swift
class CollectionViewController: UICollectionViewController {
      override func collectionView(
              _ collectionView: UICollectionView,
              cellForItemAt indexPath: IndexPath
          ) -> UICollectionViewCell {
              let cell = collectionView.dequeueReusableCell(
                  withClass: ProductCollectionViewCell.self,
                  for: indexPath
              )
              
              guard let product = initialProductsListPage?.pages[safe: indexPath.item] else {
                  return cell
              }
              
              let indexPathForCell = collectionView.indexPath(for: cell) // ⛔️ indexPathForCell이 nil이 반환됨
              cell.configureCollectionContent(with: product, indexPath: indexPath, indexPathForCell: indexPathForCell)
              cell.configureCollectionCellLayer()
              
              return cell
          }
}

class ProductCollectionViewCell: UICollectionViewCell {
      func configureCollectionContent(with product: Product, indexPath: IndexPath, indexPathForCell: IndexPath?) {
              DispatchQueue.main.async {
                  guard indexPath == indexPathForCell else { return }
                  self.productThumbnail?.image = self.getImage(from: product.thumbnail)
              }
              productName?.attributedText = product.attributedName
              productPrice?.attributedText = product.attributedPrice
              productStock?.attributedText = product.attributedStock
          }
}
```

Jager-yoo committed:
feat: ImageCache 기능 구현

Jager-yoo committed:
refactor: 이미지를 캐싱을 reloadData() 메서드가 불리기 전에 하도록 수정

- NSCache 프로퍼티를 싱글톤으로 갖는 ImageCacheManager 타입 구현
- forEach 반복문을 돌며 이미지 캐싱만 담당하는 downloadImage 메서드 구현

yeahg-dev committed:
refactor: url로부터 이미지를 비동기로 다운 받아 UIImage에 적용시키는 메서드를 UIImageView+Extension에 구현

Jager-yoo committed:
refactor: ProductCollectionViewCell 의 layer 속성 변경을 awakeFromNib 메서드에서 하도록 수정

- 기존에는 cell 의 layer 속성 변경을 뷰컨의 cellForRowAt 메서드에서 각각 해주는 방식이었음
- awakeFromNib 메서드에서 layer 를 미리 바꿔서 cell 을 생성하면, cellForRowAt 메서드 내에서는 UI 작업을 최소화할 수 있음
- 최종적으로 Main thread 의 부하를 줄일 수 있음

yeahg-dev committed:
fix: cell이 재사용되면서 비동기dataTask가 쌓여서 이미지가 변경되는 이슈 해결

Jager-yoo committed:
fix: invalidImage 의 tintColor 가 바뀌지 않는 이슈 해결

yeahg-dev committed:
feat: TableView Pagination 기능 구현

Jager-yoo committed:
feat: CollectionView Pagination 기능 구현

yeahg-dev committed:
style: 메서드 줄바꿈 수정

yeahg-dev committed:
feat: 첫 화면 로딩시 띄워줄 loadingIndicator 구현

Jager-yoo committed:
refactor: 상속 가능성이 없는 class 타입에 final 키워드 부여

kcharliek commented:
### collectionView.indexPath(for: cell) 에 대해

cell을 indexPath로 만들었는데요.
collectionView로부터 cell의 indexPath를 다시 얻어낼 필요가 있을까요? cell의 indexPath는 DataSource가 준 indexPath와 같습니다. 

그 외 `cell의 이미지가 바뀌는 문제` 에 대해서는 음성으로 설명드렸습니다. :)

yeahg-dev committed:
file: UITableView, UICollectionView의 익스텐션을 Extension 폴더로 이동

- dequeueReusableCell 메서드는 비지니스 로직이 없는 순수 유틸로 기능하므로 Extension file로 분리함

Jager-yoo committed:
refactor: Data 타입 extension 의 append 메서드 내부에서 사용되는 String.Encoding 타입을 파라미터로 분리

yeahg-dev committed:
fix: constraint 충돌 이슈 해결 #1

Jager-yoo committed:
fix: UILabel 글자가 잘리는 이슈 해결 #2

yeahg-dev committed:
UICollectionViewCell의 UILabel 추가 및 오토레이아웃 조정

- discountedPrice 레이블을 추가 후 Price레이블과 스택뷰로 묶음
- Cell내의 모든 UI컴포넌트를 단일 스택뷰로 묶음

Jager-yoo committed:
refactor: 가격을 보여주는 UILabel 을 2개로 변경, 오토레이아웃 조정 #3

- 기존에는 하나의 UILabel 에 정상 가격과 할인 가격을 모두 보여줬으나, 컬렉션뷰셀에서 가격 간의 '줄바꿈'을 구현하기 위해 UILabel 을 1개 추가함
- 할인을 하지 않으면, 가격 UILabel 이 중앙정렬될 수 있도록 오토레이아웃 조정함
- *이슈* 기기마다 일관성있는 레이아웃을 보여주기 위해, 컬렉션뷰셀의 height 값을 intrinsic size 로 줄 수 있는 방법을 찾아야 함

yeahg-dev committed:
fix: 컬렉션뷰 셀에서 가격레이블 잘리는 이슈 해결

Jager-yoo commented:
찰리 안녕하세요! 벌써 자정이 넘었네요 😢 

오늘 하루종일 릴리와 열심히 리팩토링 진행했습니다!
이제는 머지 요청드리고자 합니다! 🥰 이번 스텝에서 정말 많은 것들을 배우고 있습니다. 다음 스텝에서도 열심히 해볼게요!!

## ✅ 스텝2 관련 추가/수정한 부분
- `이미지 캐싱` 기능 추가
- 첫 화면이 보여지기 전에 `로딩 인디케이터` 추가
- `페이지네이션` 기능 추가
- 콘솔로그에 찍히던 `오토레이아웃` 관련 이슈 해결
- 컬렉션뷰셀에서 가격이 좌/우가 아니라, 위/아래로 배치되도록 레이아웃 조정

## 🩹 다음 스텝에서 더 해결할 부분
- 상품 리스트를 다운받는 데에 실패하면(서버에 문제가 있으면) 어떻게 사용자 친화적으로 에러 처리할지 고민해보기
- 2개의 뷰컨(테이블, 컬렉션)에서 공통적으로 사용되는 메서드를 어떻게 `재사용`할지 고민해보기
- 컬렉션뷰셀은 `Dynamic Type` 이 적용되지 않는 이슈 해결하기 (테이블뷰에선 잘 되고, 동일한 연산 프로퍼티로 값을 할당하고 있음)
- 컬렉션뷰셀의 `height` 가 다양한 기기(화면)에 대응할 수 있도록, `intrinsic size` 를 갖도록 만들기
  - 위 내용과 관련해서, 저희가 이슈로 작성해둔 링크를 첨부합니다! (시간 나실 때 한 번 읽어주시면 감사하겠습니다.. 🙏🏻 )
  - https://github.com/Jager-yoo/ios-open-market/issues/3

kcharliek commented:
고생많았습니다. 이슈쪽에 코멘트 달아두었으니 참고부탁드려요 :)
